Updates URDF and SDF Parsers to Support Relative Paths to Models That Contain package:// References

liangfok opened the pull request:
This PR:
1. Adds a to string operator for `PackageMap`, which is useful for debugging purposes.
2. Modifies `compareRigidBodySystems` to support relative-path command line model file arguments, which fixes the problem that resulted in #4568 and re-enables the functionality in #4516.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4576)
<!-- Reviewable:end -->

liangfok commented:
+@russtedrake for feature review please since this should get the multi-car-multi-host demo (originally introduced in #4516) working again.
+@sammy-tri for platform review please.

Thanks!

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZXwGeuGpoQQTf_kyJP:b-cuvo3r)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZY-Bp46gQoLVeL_N1u:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+@jwnimmer-tri 

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/parsers/package_map.h, line 65 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0CQruMgVU6nnijiu:-KZY0CQsdR7wtfhgv8et:buejvad) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/parsers/package_map.h#L65)):*
> ```objc
>   void PopulateUpstreamToDrake(const std::string& model_file);
> 
>   friend std::ostream& operator<<(std::ostream& out, const PackageMap& point);
> ```

BTW `point` is weird here.

---

*[drake/multibody/parsers/test/package_map_test/package_map_test.cc, line 116 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY15jK6htduCmbV8Hw:-KZY15jK6htduCmbV8Hx:b-sxcsad) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/parsers/test/package_map_test/package_map_test.cc#L116)):*
> ```cpp
> 
>   PackageMap package_map;
>   for (auto const& it : expected_packages) {
> ```

BTW META `const auto&`

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0y79gBP-PdsQ30Va:-KZY0y79gBP-PdsQ30Vb:b-d8xa9v) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L30)):*
> ```cpp
> 
> // Given a file that is assumed to be in the current working directory, this
> // method obtains the full path to the file.
> ```

BTW The method tries and sometimes does this.  But sometimes it doesn't.

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0ms4G4vxFwoUkOlm:-KZY0ms4G4vxFwoUkOln:bv1mxo5) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L34)):*
> ```cpp
>   std::string result = model_file;
>   char cwd[1024];
>   if (getcwd(cwd, sizeof(cwd)) != NULL) {
> ```

BTW What about `spruce::setAsCurrent` instead?

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0HCvhWkjv7KqHnEI:-KZY0HCvhWkjv7KqHnEJ:b-tdtvqr) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L47)):*
> ```cpp
>   // Creates a rigid body system using the first model.
>   auto r1 = make_shared<RigidBodySystem>();
>   r1->AddModelInstanceFromFile(GetFullPath(model_file_1), kQuaternion);
> ```

Why is it appropriate for this test to perform the `cwd` logic, instead of the (eventual file-loading implementation that underlies) `AddModelInstanceFromFile`.  At first glance, it seems like every program loading a file would desire this feature.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZY1EamMRbvJV6U810M:bv95ifn)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/parsers/package_map.h, line 65 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0CQruMgVU6nnijiu:-KZY913fryToJhkXILdA:b-yxcijg) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/parsers/package_map.h#L65)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW `point` is weird here.
</blockquote></details>

Oops! Fixed, thanks.

---

*[drake/multibody/parsers/test/package_map_test/package_map_test.cc, line 116 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY15jK6htduCmbV8Hw:-KZY9LeRQG7malhLtIex:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/parsers/test/package_map_test/package_map_test.cc#L116)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW META `const auto&`
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0y79gBP-PdsQ30Va:-KZY9lwhDKEa2caCR4gV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L30)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW The method tries and sometimes does this.  But sometimes it doesn't.
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0ms4G4vxFwoUkOlm:-KZY9pTWsSrsvK6g3Ttf:bdew2nd) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L34)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW What about `spruce::setAsCurrent` instead?
</blockquote></details>

I'm afraid that might modify the current working directory

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0HCvhWkjv7KqHnEI:-KZYBbQjIaQnEb643-CF:b7r235y) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L47)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Why is it appropriate for this test to perform the `cwd` logic, instead of the (eventual file-loading implementation that underlies) `AddModelInstanceFromFile`.  At first glance, it seems like every program loading a file would desire this feature.
</blockquote></details>

I believe all existing calls to `AddFooFromFile()` provide the full path to the file, except for this unit test. How about let's wait for the rule-of-three to apply before moving this logic into a more general location? I added a TODO comment and updated the documentation for the parsers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending the discussion w jeremy. Fwiw,  I think a simpler solution might have been to the specific prius problem might have been to add the oackage to the urdf file, too?

---

Reviewed 3 of 6 files at r1, 4 of 5 files at r2.
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 29 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYH7MuGnxrrqHiBwQl:-KZYH7MuGnxrrqHiBwQm:b-jnlbqr) ([raw file](https://github.com/robotlocomotion/drake/blob/bca81daccda08bf1b1a8e64b9a4e0e383475ed71/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L29)):*
> ```cpp
> 
> // TODO(liang.fok): Once the rule-of-three is met, move this into a more general
> // location so it can be shared.
> ```

BTW, I don't think there is any question that the rule of three will be met on this soon.  Would also favor pulling it out sooner.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZYHSjeJCVEL-VFiYS_:b-m9hh8m)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Reviewed 1 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

liangfok commented:
> I think a simpler solution might have been to the specific prius problem might have been to add the oackage to the urdf file, too?

Do you mean modifying `CMakeLists.txt` to contain the following?

```
drake_add_test(NAME comparePriusURDFtoSDF COMMAND compareRigidBodySystems package://prius/prius.sdf package://prius/prius.urdf WORKING_DIRECTORY "${PROJECT_SOURCE_DIR}/automotive/models/prius")
```
That unfortunately will not work due to the explanation here: https://github.com/RobotLocomotion/drake/pull/4516#issuecomment-268567686.

---

Review status: 7 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 29 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYH7MuGnxrrqHiBwQl:-KZYb9o7QbIVd3nMqu3h:b-fig8rk) ([raw file](https://github.com/robotlocomotion/drake/blob/bca81daccda08bf1b1a8e64b9a4e0e383475ed71/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L29)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

BTW, I don't think there is any question that the rule of three will be met on this soon.  Would also favor pulling it out sooner.
</blockquote></details>

Done. I moved it to `parser_common.h`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZYKLxw0buhmVcwqkaa:brsh442)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Checkpoint.

---

Reviewed 4 of 5 files at r2, 1 of 3 files at r3.
Review status: 8 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parsers/sdf_parser.h, line 28 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYe67bQ6t1L0Ggs2Ly:-KZYe67bQ6t1L0Ggs2Lz:b2ycavs) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/sdf_parser.h#L28)):*
> ```objc
>  * AddModelInstancesFromSdfFileToWorldSearchingInRosPackages().
>  *
>  * @param[in] filename The full path to the SDF file containing the model to be
> ```

META for both parsers: why is this a change that users would want?

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0ms4G4vxFwoUkOlm:-KZYbihwOi_0b9QBE9Da:bunyjjs) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L34)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I'm afraid that might modify the current working directory
</blockquote></details>

Why?  It's a platform-agnostic wrapper for `getcwd`, and already used by the parser: https://github.com/RobotLocomotion/drake/blob/cff68749f092712a73377aff4ca613a20af37dee/drake/multibody/parsers/parser_common.cc#L93

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0HCvhWkjv7KqHnEI:-KZYcVEvUeCTBlRLstEf:b-fstqk3) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L47)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I believe all existing calls to `AddFooFromFile()` provide the full path to the file, except for this unit test. How about let's wait for the rule-of-three to apply before moving this logic into a more general location? I added a TODO comment and updated the documentation for the parsers.
</blockquote></details>

I don't think having GetFullPath in a common library is any better.  Let me try to restate...

We pass a (possibly relative) filename to the RBS, which then just gives it to a parser.  In other cases or other code, we give a (possibly relative) filename to the parser directly.

For any case that matters in this discussion, the parser can open that fie.  There is a `path/to/foo.sdf` or on disk, relative to the cwd.  Sometimes, the path was already `/absolute/path/to/bar.sdf`.

Relatedly, we have a heuristic that if I am opening some `foo.sdf`, then I look in the parent directories of `foo.sdf` to find the most-nearby`package.xml` file, and use that for loading the package resources within the sdf.

My point is that, given that we can always find and open the SDF, why can't we always automatically find the package.xml also?  Why we should impose upon calling code the burden to absoluteify their paths before calling the parser?

In short, the parser can call GetFullPath internally, and use that to populate the package map.  Given that's possible, why is the current PR's approach superior to that approach?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZYeOTd4g-e5Ij-BOVn:by96gdd)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm_cancel:" alt=":lgtm_cancel:" align="absmiddle" src="https://reviewable.io/lgtm_cancel.png" height="20" width="61"/> (nothing in particular but enough has changed I want to review again tomorrow when I have time)

---

Review status: 8 of 10 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZYewHQrgeu0JvQF0C3:b-b7b7rg)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 8 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/parsers/sdf_parser.h, line 28 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYe67bQ6t1L0Ggs2Ly:-KZYhYSATg3VYt2AOdZS:b-ayuggh) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/sdf_parser.h#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

META for both parsers: why is this a change that users would want?
</blockquote></details>

It adds clarity to what was actually being supported. The fact that relative paths used to work in some situations was, in my opinion, a coincident due to TinyXML's use of `fopen()`.

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0ms4G4vxFwoUkOlm:-KZYiSkjQnZg8n-liiAB:be76bs2) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L34)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Why?  It's a platform-agnostic wrapper for `getcwd`, and already used by the parser: https://github.com/RobotLocomotion/drake/blob/cff68749f092712a73377aff4ca613a20af37dee/drake/multibody/parsers/parser_common.cc#L93
</blockquote></details>

Yes, you're right. I've switched to use it in R3.

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0HCvhWkjv7KqHnEI:-KZYihSNOVbMz4h0y5ct:b-a1rldi) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L47)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't think having GetFullPath in a common library is any better.  Let me try to restate...

We pass a (possibly relative) filename to the RBS, which then just gives it to a parser.  In other cases or other code, we give a (possibly relative) filename to the parser directly.

For any case that matters in this discussion, the parser can open that fie.  There is a `path/to/foo.sdf` or on disk, relative to the cwd.  Sometimes, the path was already `/absolute/path/to/bar.sdf`.

Relatedly, we have a heuristic that if I am opening some `foo.sdf`, then I look in the parent directories of `foo.sdf` to find the most-nearby`package.xml` file, and use that for loading the package resources within the sdf.

My point is that, given that we can always find and open the SDF, why can't we always automatically find the package.xml also?  Why we should impose upon calling code the burden to absoluteify their paths before calling the parser?

In short, the parser can call GetFullPath internally, and use that to populate the package map.  Given that's possible, why is the current PR's approach superior to that approach?
</blockquote></details>

Yes, you're right that having the parsers internally derive the full path may be an improvement this PR, but I would argue that this PR is still a strict improvement over what's in master, which happens to work because of some internal details within TinyXML. Let me investigate this further to see if it's possible to move the relative-path-to-full-path logic within the parsers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 8 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/parsers/sdf_parser.h, line 28 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYe67bQ6t1L0Ggs2Ly:-KZYtMbqkW0ldzxUS8eb:bnaia7r) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/sdf_parser.h#L28)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

It adds clarity to what was actually being supported. The fact that relative paths used to work in some situations was, in my opinion, a coincident due to TinyXML's use of `fopen()`.
</blockquote></details>

I don't understand how `fopen` or TinyXML's use of it relates to anything?  We ask TinyXML to `LoadFile`, giving it a filename string.  If the filename denotes a file that exists, it opens the file; if not, it errors.  Passing a relative filename string and having it be interpreted with respect to the cwd is an obvious feature of, like, all file-open APIs ever?

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0HCvhWkjv7KqHnEI:-KZYuIvAQisNeaeThhL_:b7ihalg) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L47)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Yes, you're right that having the parsers internally derive the full path may be an improvement this PR, but I would argue that this PR is still a strict improvement over what's in master, which happens to work because of some internal details within TinyXML. Let me investigate this further to see if it's possible to move the relative-path-to-full-path logic within the parsers.
</blockquote></details>

I could probably buy the argument "We really need to edit the `prius.sdf`s to use `package://` for their mesh because reasons, and we need to merge that _now now now_" -- if accompanied by appropriate `TODO`s to fix the C++ code.

I still don't understand how TinyXML related to any part of that discussion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 3 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/parsers/parser_common.cc, line 12 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYwWVqqZE4gBmHVx1q:-KZYwWVqqZE4gBmHVx1r:baw2eac) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/parser_common.cc#L12)):*
> ```cpp
> #include "drake/multibody/joints/fixed_joint.h"
> #include "drake/multibody/joints/floating_base_types.h"
> #include "drake/thirdParty/bsd/spruce/spruce.hh"
> ```

The prior path spelling was correct ; this new spelling is incorrect.

---

*[drake/multibody/parsers/parser_common.cc, line 33 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYwsdW2jyOl_C7mAeh:-KZYwsdW2jyOl_C7mAei:b919foi) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/parser_common.cc#L33)):*
> ```cpp
>   path.append(file_name);
>   if (path.isFile()) {
>     result = path.getStr();
> ```

Missing test coverage.  I don't system1 model-compare regression test counts -- something should test this within the parsers subdirectory at least.

---

*[drake/multibody/parsers/parser_common.h, line 11 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYw61k5YXLRzL1tj10:-KZYw61k5YXLRzL1tj11:b-yrumkl) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/parser_common.h#L11)):*
> ```objc
> #include "drake/multibody/rigid_body_frame.h"
> #include "drake/multibody/rigid_body_tree.h"
> #include "drake/thirdParty/bsd/spruce/spruce.hh"
> ```

Unused

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/parsers/sdf_parser.h, line 28 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYe67bQ6t1L0Ggs2Ly:-KZZ-qZvuug7DAQQ1AS_:bp50v62) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/sdf_parser.h#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't understand how `fopen` or TinyXML's use of it relates to anything?  We ask TinyXML to `LoadFile`, giving it a filename string.  If the filename denotes a file that exists, it opens the file; if not, it errors.  Passing a relative filename string and having it be interpreted with respect to the cwd is an obvious feature of, like, all file-open APIs ever?
</blockquote></details>

When we pass TinyXML a relative-path file name, it happens to work when the executable is launched from a directory that happens to complete the relative-path file name. This, in my opinion, was coincidental. The API, as I had originally intended, required absolute path information. (I will now see if I can relax this requirement by using `GetFullPath()` within the parsers.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/multibody/parsers/sdf_parser.h, line 28 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYe67bQ6t1L0Ggs2Ly:-KZZ1JK_TWksWeGCZc8I:b4ar5b3) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/sdf_parser.h#L28)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

When we pass TinyXML a relative-path file name, it happens to work when the executable is launched from a directory that happens to complete the relative-path file name. This, in my opinion, was coincidental. The API, as I had originally intended, required absolute path information. (I will now see if I can relax this requirement by using `GetFullPath()` within the parsers.)
</blockquote></details>

Support for passing a relative pathname to an SDF is how Drake has worked since ~forever, TTBOMK.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 4 of 5 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/parsers/sdf_parser.h, line 28 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYe67bQ6t1L0Ggs2Ly:-KZb5oyuK2BKsnASPtG8:bnr4c97) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/sdf_parser.h#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Support for passing a relative pathname to an SDF is how Drake has worked since ~forever, TTBOMK.
</blockquote></details>

BTW IMO when you pass any library/tool a relative path name it always happens to work iff the executable is launched with the appropriate cwd.  I'm not sure why this case would be different.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
From #4516 -> here (since the discussion is about this PR).  

"There is one scenario that currently doesn't work, which #4576 will fix: We currently cannot load a model file using a relative path when the model file contains references to mesh files using package:// syntax. This is because the parsers are currently designed to crawl up the directory tree from the file name and fail to do this when it is provided a relative path."

I cannot see how this PR changes this behavior, so I suspect I must be missing something here.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZbCk6nk4BBI1fp5XAW:b8zrc2u)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@sammy-tri: Sorry I should have applied the "Do Not Review" label to this PR while I attempt to move the `GetFullPath` functionality into the parsers.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZbGPaVMR9SMtI4bXgl:bgi3jfo)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 5 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/parsers/parser_common.cc, line 12 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYwWVqqZE4gBmHVx1q:-KZtIBH_DYQ5IezRQagB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/parser_common.cc#L12)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The prior path spelling was correct ; this new spelling is incorrect.
</blockquote></details>

Done.

---

*[drake/multibody/parsers/parser_common.cc, line 33 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYwsdW2jyOl_C7mAeh:-KZtIFtPRMhu45Q-TB1I:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/parser_common.cc#L33)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Missing test coverage.  I don't system1 model-compare regression test counts -- something should test this within the parsers subdirectory at least.
</blockquote></details>

Done.

---

*[drake/multibody/parsers/parser_common.h, line 11 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYw61k5YXLRzL1tj10:-KZtIQbUKvtmGg5PYevK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/parser_common.h#L11)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unused
</blockquote></details>

Done.

---

*[drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZY0HCvhWkjv7KqHnEI:-KZtI_s4CRVsNFqTDWlM:b-k2uaag) ([raw file](https://github.com/robotlocomotion/drake/blob/0af1372ac8a9115c792329421ad4c2482e17f41d/drake/multibody/rigid_body_system1/test/compareRigidBodySystems.cpp#L47)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I could probably buy the argument "We really need to edit the `prius.sdf`s to use `package://` for their mesh because reasons, and we need to merge that _now now now_" -- if accompanied by appropriate `TODO`s to fix the C++ code.

I still don't understand how TinyXML related to any part of that discussion.
</blockquote></details>

Done -- parsers now compute full path given relative path.

---

*[drake/multibody/parsers/sdf_parser.h, line 28 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZYe67bQ6t1L0Ggs2Ly:-KZtISqqwKyVyvaDU8aM:byok759) ([raw file](https://github.com/robotlocomotion/drake/blob/6e12b0d279f516d99dd45ff5e0c575808baefcef/drake/multibody/parsers/sdf_parser.h#L28)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW IMO when you pass any library/tool a relative path name it always happens to work iff the executable is launched with the appropriate cwd.  I'm not sure why this case would be different.
</blockquote></details>

Done. Relative paths should now be supported by the parsers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZtO1V9KEYA9S2M1ieI:bblnx7v)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 12 of 12 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/parsers/parser_common.cc, line 30 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZvMK_nXHGwD6whx2g-:-KZvMK_nXHGwD6whx2g0:b90uper) ([raw file](https://github.com/robotlocomotion/drake/blob/3697f35425d1a33c379359d77811b5dc362d1b77/drake/multibody/parsers/parser_common.cc#L30)):*
> ```cpp
> std::string GetFullPath(const std::string& file_name) {
>   std::string result = file_name;
>   if (result.size() == 0) {
> ```

BTW `.empty()` if you wish.

---

*[drake/multibody/parsers/parser_common.cc, line 36 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZvUkhDx3-6NPLE93w1:-KZvUkhELs03QNNlYtQ6:b9coct0) ([raw file](https://github.com/robotlocomotion/drake/blob/3697f35425d1a33c379359d77811b5dc362d1b77/drake/multibody/parsers/parser_common.cc#L36)):*
> ```cpp
> 
>   const std::string prefix = "/";
>   if (result.substr(0, prefix.size()) == prefix) {
> ```

BTW I was surprised that spruce doesn't offer this as a predicate, but it does not.  Wow!

---

*[drake/multibody/parsers/parser_common.cc, line 45 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZvVDMCn_Iq61PhNyeK:-KZvVDMDCGZw0wrzVzLx:b-lxh7p1) ([raw file](https://github.com/robotlocomotion/drake/blob/3697f35425d1a33c379359d77811b5dc362d1b77/drake/multibody/parsers/parser_common.cc#L45)):*
> ```cpp
>     }
>     if (!path.exists()) {
>       throw std::runtime_error("drake::parsers::GetFullPath: ERROR: "
> ```

This branch seems to be unreachable.  Inside spruce, `isFile` checks `exists` first.

---

*[drake/multibody/parsers/parser_common.cc, line 54 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZvVMeAGuusIbJTL5eH:-KZvVMeBOyUEzrpPrCcC:b-bij3db) ([raw file](https://github.com/robotlocomotion/drake/blob/3697f35425d1a33c379359d77811b5dc362d1b77/drake/multibody/parsers/parser_common.cc#L54)):*
> ```cpp
>     path.setAsCurrent();
>     path.append(file_name);
>     if (path.isFile() && path.exists()) {
> ```

The second element of the conjunction is redundant.

---

*[drake/multibody/parsers/test/parser_common_test/parser_common_test.cc, line 65 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZvWnhfRPuEuGKXpImb:-KZvWnhgrym8-xCbxkbh:b-4s781x) ([raw file](https://github.com/robotlocomotion/drake/blob/3697f35425d1a33c379359d77811b5dc362d1b77/drake/multibody/parsers/test/parser_common_test/parser_common_test.cc#L65)):*
> ```cpp
>   string full_path;
>   full_path = GetFullPath(original_path);
>   // ASSERT_NO_THROW(full_path = GetFullPath(original_path));
> ```

Commented-out code.

---

*[drake/multibody/parsers/test/parser_common_test/test_file.txt, line 0 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-KZvWIUDPJhBI163_kAe:-KZvWIUDPJhBI163_kAf:bl03qwk) ([raw file](https://github.com/robotlocomotion/drake/blob/3697f35425d1a33c379359d77811b5dc362d1b77/drake/multibody/parsers/test/parser_common_test/test_file.txt#L0)):*
BTW META Consider adding a comment to empty files, to prevent any developer from being confused whether the file is empty on purpose, or due to some kind of mistake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
> updated the BUILD file for Bazel support.

Missed git add or push?  I'm not seeing it (though it's fine to omit if you wish).

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZxbwCkoulIkV2gN4BL:b7zhgkz)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-KZxc4c0RR77C4CiEjpK:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please

liangfok commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please

jwnimmer-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 12 files at r4, 5 of 6 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-:-K_-wbRrIYqJRSgTp--g:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 5 of 6 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/parsers/package_map.cc, line 132 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-K_-vke5Pu8TK9_3bcKk:-K_-vke5Pu8TK9_3bcKl:b-a5je3i) ([raw file](https://github.com/robotlocomotion/drake/blob/38aa1c81d550e23a7dd86d64a2fe72caa539af60/drake/multibody/parsers/package_map.cc#L132)):*
> ```cpp
> 
> // Returns true if and only if string @p a ends with string @p b.
> bool StringEndsWith(const std::string& a, const std::string& b) {
> ```

Helpers should have private linkage; so either in `namespace {` anoymous, or marked `static`.

---

*[drake/multibody/parsers/package_map.cc, line 136 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-K_-wCoP4wxlooDjAkTJ:-K_-wCoQ9-lEC0R_BVUV:b-8t6sqq) ([raw file](https://github.com/robotlocomotion/drake/blob/38aa1c81d550e23a7dd86d64a2fe72caa539af60/drake/multibody/parsers/package_map.cc#L136)):*
> ```cpp
>   return std::equal(a.begin() + a.size() - b.size(), a.end(), b.begin());
> }
> void PackageMap::PopulateUpstreamToDrakeHelper(const string& directory) {
> ```

BTW Consider blank line between methods?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 15 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/parsers/package_map.cc, line 132 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-K_-vke5Pu8TK9_3bcKk:-K_-xlt33w6cXpVrxVuY:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/38aa1c81d550e23a7dd86d64a2fe72caa539af60/drake/multibody/parsers/package_map.cc#L132)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Helpers should have private linkage; so either in `namespace {` anoymous, or marked `static`.
</blockquote></details>

Done. Thanks.

---

*[drake/multibody/parsers/package_map.cc, line 136 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4576#-K_-wCoP4wxlooDjAkTJ:-K_-xn1SOfVY5HCUTZEi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/38aa1c81d550e23a7dd86d64a2fe72caa539af60/drake/multibody/parsers/package_map.cc#L136)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Consider blank line between methods?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please

jwnimmer-tri commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4576)*
<!-- Sent from Reviewable.io -->
